core/src/main/java/org/elasticsearch/index/query/BoolQueryParser.java

```diff
@@ -71,14 +71,12 @@ public QueryBuilder fromXContent(QueryParseContext parseContext) throws IOExcept
                 switch (currentFieldName) {
                 case "must":
                     query = parseContext.parseInnerQueryBuilder();
-                    if (query != null) {
-                        mustClauses.add(query);
-                    }
+                    mustClauses.add(query);
                     break;
                 case "should":
                     query = parseContext.parseInnerQueryBuilder();
-                    if (query != null) {
-                        shouldClauses.add(query);
+                    shouldClauses.add(query);
+                    if (query != EmptyQueryBuilder.PROTOTYPE) {
                         if (parseContext.isFilter() && minimumShouldMatch == null) {
```

Comment:
I think i am missing something here, but when can isFilter be true here? I also wonder if the conditional here is really required or we could drop it?